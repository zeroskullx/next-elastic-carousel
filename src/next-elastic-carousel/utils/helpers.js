export const noop = () => {};

export const numberToArray = n => [...Array(n).keys()];

export const cssPrefix = (...classNames) => {
  const prefix = "rec";
  const space = " ";
  let result = `${prefix}`; // initial it with global prefix;

  // in case of an array we add the class prefix per item;
  const chainedClasses = classNames.reduce((acc, current) => {
    if (current) {
      acc += `${space}${prefix}-${current}`; // we must keep spaces between class names
    }
    return acc;
  }, "");
  result += chainedClasses;

  return result;
};

export const pipe = (...fns) => x => fns.reduce((v, f) => f(v), x);

export const throttle = (func, limit) => {
  let lastFunc;
  let lastRan;
  return function() {
    const context = this;
    const args = arguments;
    if (!lastRan) {
      func.apply(context, args);
      lastRan = Date.now();
    } else {
      clearTimeout(lastFunc);
      lastFunc = setTimeout(function() {
        if (Date.now() - lastRan >= limit) {
          func.apply(context, args);
          lastRan = Date.now();
        }
      }, limit - (Date.now() - lastRan));
    }
  };
};
