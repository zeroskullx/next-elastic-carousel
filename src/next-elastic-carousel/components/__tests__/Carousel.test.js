import * as React from 'react'
import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'

import Carousel from '../Carousel'
import { numberToArray } from '../../utils/helpers'
import { snapshot } from '../../test'

describe('Carousel - public API (props)', () => {
  const Items = numberToArray(5).map((i) => (
    <div className="test-child" key={i}>
      {i}
    </div>
  ))

  snapshot(<Carousel>{Items}</Carousel>, 'default')

  it('renders without crashing', () => {
    render(<Carousel>{Items}</Carousel>)
  })

  it('renders children', () => {
    const { container } = render(<Carousel>{Items}</Carousel>)
    const children = container.querySelectorAll('.test-child') // Usando querySelectorAll aqui
    expect(children.length).toBe(Items.length)
  })

  it("one child won't break on next", () => {
    // Renderizando o componente com apenas um item
    const { container } = render(<Carousel>{Items[0]}</Carousel>)

    // Encontrando o botão "next" usando a classe "rec-arrow-right"
    const nextButton = container.querySelector('button.rec-arrow-button-right')

    // Verifique se o botão foi encontrado
    expect(nextButton).toBeInTheDocument()

    // Simulando o clique no botão "next"
    fireEvent.click(nextButton)

    // Adicione uma asserção aqui para garantir que o comportamento esperado ocorre após o clique
    // Por exemplo: verificar se a navegação ocorreu corretamente
    expect(nextButton).toBeInTheDocument() // Pode ser ajustado conforme o comportamento esperado
  })

  it('renders with className in root', () => {
    const testClassName = 'test-root'
    const { container } = render(
      <Carousel className={testClassName}>{Items}</Carousel>
    )
    const carousel = container.querySelector(`.${testClassName}`) // Usando querySelector aqui
    expect(carousel).toHaveClass(testClassName)
  })

  it('renders with style in root', () => {
    const styleToRender = { position: 'fixed' }
    const { container } = render(
      <Carousel style={styleToRender}>{Items}</Carousel>
    )

    // Pega o elemento raiz do container
    const carousel = container.firstChild

    // Verifica se o estilo foi aplicado corretamente
    expect(carousel).toHaveStyle('position: fixed')
  })

  it('renders with verticalMode set to true', () => {
    const { container } = render(
      <Carousel verticalMode={true}>{Items}</Carousel>
    )
    const slider = container.querySelector(`.rec-slider`)
    expect(slider).toHaveStyle('flex-direction: column')
  })

  it('isRTL prop is passed to Slider and styles are applied correctly', () => {
    // Teste para isRTL = true
    const { container: containerRTL } = render(
      <Carousel isRTL={true}>{Items}</Carousel>
    )
    const sliderRTL = containerRTL.querySelector(`.rec-slider`)

    expect(sliderRTL).toHaveStyle({
      left: 'auto',
      right: '0px',
    })

    // Teste para isRTL = false
    const { container: containerLTR } = render(
      <Carousel isRTL={false}>{Items}</Carousel>
    )
    const sliderLTR = containerLTR.querySelector(`.rec-slider`)

    expect(sliderLTR).toHaveStyle({
      left: '0px',
      right: 'auto',
    })
  })

  it('wont break with outerSpacing', () => {
    const { container } = render(
      <Carousel outerSpacing={100}>{Items}</Carousel>
    )
    const carousel = container.querySelector('.rec-carousel')
    expect(carousel).toBeInTheDocument()
  })

  it('renderPagination (renders custom pagination)', () => {
    const CustomPagination = () => <div>Render Pagination</div>
    const renderPagination = () => <CustomPagination />

    let componentInstance = null

    // Renders the component and captures the instance
    React.act(() => {
      render(
        <Carousel
          ref={(instance) => {
            componentInstance = instance
          }}
          renderPagination={renderPagination}
        />
      )
    })

    //console.log(componentInstance)

    // Checks if the instance was captured
    expect(componentInstance).not.toBeNull()

    // Updates the state directly
    React.act(() => {
      componentInstance.setState({ rootHeight: 100 })
    })

    // Checks the updated state
    expect(componentInstance.state.rootHeight).toBe(100)

    expect(screen.getByText('Render Pagination')).toBeInTheDocument()
  })
})

describe('Carousel - public CSS classnames', () => {
  const publicClasses = [
    'carousel-wrapper',
    'carousel',
    'slider-container',
    'slider',
    'carousel-item',
    'carousel-item-visible',
    'carousel-item-hidden',
    'carousel-item-prev',
    'carousel-item-next',
    'item-wrapper',
    'arrow',
    'arrow-button-right',
    'arrow-button-left',
    'swipable',
    //'dot',
    //'dot_active',
    //'pagination',
  ]
  const prefix = 'rec'
  const Items = numberToArray(5).map((i) => (
    <div className="test-child" key={i}>
      {i}
    </div>
  ))

  publicClasses.forEach((className) => {
    const withPrefix = `${prefix}-${className}`

    it(`renders ${withPrefix}`, () => {
      const { container } = render(
        <Carousel
          pagination={true}
          initialActiveIndex={1}
          itemsToShow={1}
          enableSwipe={true}
        >
          {Items}
        </Carousel>
      )

      const elementWithClass = container.querySelector(`.${withPrefix}`)

      expect(elementWithClass).toBeInTheDocument()
    })
  })
})
