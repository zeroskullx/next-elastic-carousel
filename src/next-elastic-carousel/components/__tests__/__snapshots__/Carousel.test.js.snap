// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel - public API (props) default 1`] = `
<div>
  <div
    class="rec rec-carousel-wrapper css-x33sbs"
    overflow="visible"
  >
    <div
      class="rec rec-carousel css-1mezvkn"
      opacity="1"
      overflow="visible"
    >
      <div
        class="rec rec-arrow rec-arrow-left css-1kjr7xd"
        direction="left"
      >
        <button
          class="rec rec-arrow-button-left css-76e8uw"
          disabled=""
          name="btn-prev"
          tabindex="0"
        >
          <svg
            class="rec rec-chevron-left"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15 18-6-6 6-6"
            />
          </svg>
        </button>
      </div>
      <div
        class="rec rec-slider-container css-2mrzk5"
      >
        <div
          class="rec rec-slider css-11ve2d9"
          style="transition: all 400ms ease; left: 0px; margin-left: -0px;"
        >
          <div
            class="rec rec-swipable"
            style="display: flex; flex-direction: row;"
          >
            <div
              class="rec rec-carousel-item rec-carousel-item-0 rec-carousel-item-visible"
              style="padding-left: 0px;"
            >
              <div
                class="rec rec-item-wrapper css-1db75p4"
                style="width: 0px;"
              >
                <div
                  class="test-child"
                  tabindex="0"
                >
                  0
                </div>
              </div>
            </div>
            <div
              class="rec rec-carousel-item rec-carousel-item-1 rec-carousel-item-hidden rec-carousel-item-next"
              style="padding-left: 0px;"
            >
              <div
                class="rec rec-item-wrapper css-1db75p4"
                style="width: 0px;"
              >
                <div
                  class="test-child"
                  tabindex="-1"
                >
                  1
                </div>
              </div>
            </div>
            <div
              class="rec rec-carousel-item rec-carousel-item-2 rec-carousel-item-hidden"
              style="padding-left: 0px;"
            >
              <div
                class="rec rec-item-wrapper css-1db75p4"
                style="width: 0px;"
              >
                <div
                  class="test-child"
                  tabindex="-1"
                >
                  2
                </div>
              </div>
            </div>
            <div
              class="rec rec-carousel-item rec-carousel-item-3 rec-carousel-item-hidden"
              style="padding-left: 0px;"
            >
              <div
                class="rec rec-item-wrapper css-1db75p4"
                style="width: 0px;"
              >
                <div
                  class="test-child"
                  tabindex="-1"
                >
                  3
                </div>
              </div>
            </div>
            <div
              class="rec rec-carousel-item rec-carousel-item-4 rec-carousel-item-hidden"
              style="padding-left: 0px;"
            >
              <div
                class="rec rec-item-wrapper css-1db75p4"
                style="width: 0px;"
              >
                <div
                  class="test-child"
                  tabindex="-1"
                >
                  4
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rec rec-arrow rec-arrow-right css-1kjr7xd"
        direction="right"
      >
        <button
          class="rec rec-arrow-button-right css-12mon0g"
          name="btn-next"
          tabindex="0"
        >
          <svg
            class="rec rec-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
